File: sdks/ios/Sources/ArbitrationSDK/Core/ArbitrationDetector.swift
import Foundation
import CoreML
import NaturalLanguage
import Vision
import Combine

/// Main arbitration detection engine for iOS
@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
public final class ArbitrationDetector: ObservableObject {
    
    // MARK: - Published Properties
    @Published public private(set) var isAnalyzing = false
    @Published public private(set) var analysisProgress: Double = 0.0
    @Published public private(set) var lastAnalysisResult: ArbitrationAnalysisResult?
    
    // MARK: - Private Properties
    private let mlModel: MLModel?
    private let networkService: NetworkService
    private let storageService: StorageService
    private let configuration: SDKConfiguration
    private var analysisTask: Task<Void, Error>?
    private let analysisQueue = DispatchQueue(label: "com.arbitrationsdk.analysis", qos: .userInitiated)
    
    // MARK: - Initialization
    public init(configuration: SDKConfiguration = .default) {
        self.configuration = configuration
        self.networkService = NetworkService(configuration: configuration)
        self.storageService = StorageService()
        
        // Load Core ML model
        do {
            if let modelURL = Bundle.module.url(forResource: "ArbitrationDetectionModel", withExtension: "mlmodelc") {
                self.mlModel = try MLModel(contentsOf: modelURL)
            } else {
                self.mlModel = nil
                print("Warning: Core ML model not found. Using fallback detection.")
            }
        } catch {
            self.mlModel = nil
            print("Error loading Core ML model: \(error)")
        }
    }
    
    // MARK: - Public Methods
    
    /// Analyze text for arbitration clauses
    public func analyzeText(_ text: String) async throws -> ArbitrationAnalysisResult {
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw ArbitrationError.invalidInput("Text cannot be empty")
        }
        
        await MainActor.run {
            isAnalyzing = true
            analysisProgress = 0.0
        }
        
        defer {
            Task { @MainActor in
                isAnalyzing = false
                analysisProgress = 0.0
            }
        }
        
        do {
            let result = try await performAnalysis(text: text)
            
            // Store result locally
            try await storageService.saveAnalysisResult(result)
            
            await MainActor.run {
                lastAnalysisResult = result
            }
            
            return result
        } catch {
            throw ArbitrationError.analysisFailure(error.localizedDescription)
        }
    }
    
    /// Analyze document from URL or file path
    public func analyzeDocument(_ documentData: Data, type: DocumentType) async throws -> ArbitrationAnalysisResult {
        let extractedText = try await extractTextFromDocument(documentData, type: type)
        return try await analyzeText(extractedText)
    }
    
    /// Analyze image containing text
    public func analyzeImage(_ image: UIImage) async throws -> ArbitrationAnalysisResult {
        let extractedText = try await extractTextFromImage(image)
        return try await analyzeText(extractedText)
    }
    
    /// Cancel ongoing analysis
    public func cancelAnalysis() {
        analysisTask?.cancel()
        Task { @MainActor in
            isAnalyzing = false
            analysisProgress = 0.0
        }
    }
    
    // MARK: - Private Methods
    
    private func performAnalysis(text: String) async throws -> ArbitrationAnalysisResult {
        let analysisId = UUID().uuidString
        let startTime = Date()
        
        await updateProgress(0.1)
        
        // Step 1: Preprocessing
        let preprocessedText = preprocessText(text)
        await updateProgress(0.2)
        
        // Step 2: Feature extraction
        let features = try await extractFeatures(from: preprocessedText)
        await updateProgress(0.4)
        
        // Step 3: ML Model prediction (if available)
        var mlConfidence: Double = 0.0
        var mlResult: Bool = false
        
        if let model = mlModel {
            (mlResult, mlConfidence) = try await predictWithMLModel(model, features: features)
        }
        await updateProgress(0.6)
        
        // Step 4: Rule-based analysis
        let ruleBasedResult = performRuleBasedAnalysis(preprocessedText)
        await updateProgress(0.8)
        
        // Step 5: Combine results
        let finalResult = combineAnalysisResults(
            mlResult: mlResult,
            mlConfidence: mlConfidence,
            ruleBasedResult: ruleBasedResult,
            originalText: text,
            analysisId: analysisId
        )
        
        await updateProgress(1.0)
        
        let endTime = Date()
        finalResult.processingTime = endTime.timeIntervalSince(startTime)
        
        return finalResult
    }
    
    private func preprocessText(_ text: String) -> String {
        return text
            .lowercased()
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
    
    private func extractFeatures(from text: String) async throws -> [String: Any] {
        return try await withCheckedThrowingContinuation { continuation in
            analysisQueue.async {
                do {
                    let features: [String: Any] = [
                        "text_length": text.count,
                        "word_count": text.components(separatedBy: .whitespacesAndNewlines).count,
                        "arbitration_keywords": self.countArbitrationKeywords(in: text),
                        "legal_terms": self.countLegalTerms(in: text),
                        "sentiment_score": self.analyzeSentiment(text),
                        "complexity_score": self.calculateComplexityScore(text)
                    ]
                    continuation.resume(returning: features)
                } catch {
                    continuation.resume(throwing: error)
                }
            }
        }
    }
    
    private func predictWithMLModel(_ model: MLModel, features: [String: Any]) async throws -> (Bool, Double) {
        return try await withCheckedThrowingContinuation { continuation in
            do {
                // Convert features to ML input format
                let input = try MLDictionaryFeatureProvider(dictionary: features)
                let prediction = try model.prediction(from: input)
                
                // Extract prediction results
                if let probabilityDict = prediction.featureValue(for: "probability")?.dictionaryValue,
                   let arbitrationProbability = probabilityDict["arbitration"]?.doubleValue,
                   let classLabel = prediction.featureValue(for: "class")?.stringValue {
                    
                    let hasArbitration = classLabel == "arbitration"
                    continuation.resume(returning: (hasArbitration, arbitrationProbability))
                } else {
                    continuation.resume(returning: (false, 0.0))
                }
            } catch {
                continuation.resume(throwing: error)
            }
        }
    }
    
    private func performRuleBasedAnalysis(_ text: String) -> RuleBasedResult {
        let arbitrationKeywords = [
            "arbitration", "arbitrator", "arbitral", "binding arbitration",
            "alternative dispute resolution", "adr", "arbitration clause",
            "dispute resolution", "binding resolution", "final and binding"
        ]
        
        let exclusionPhrases = [
            "court proceedings", "litigation", "judicial review",
            "class action", "jury trial", "court of law"
        ]
        
        var keywordMatches: [KeywordMatch] = []
        var exclusionMatches: [KeywordMatch] = []
        
        // Find arbitration keywords
        for keyword in arbitrationKeywords {
            let ranges = text.ranges(of: keyword, options: .caseInsensitive)
            for range in ranges {
                let match = KeywordMatch(
                    keyword: keyword,
                    range: range,
                    context: extractContext(from: text, around: range),
                    confidence: calculateKeywordConfidence(keyword)
                )
                keywordMatches.append(match)
            }
        }
        
        // Find exclusion phrases
        for phrase in exclusionPhrases {
            let ranges = text.ranges(of: phrase, options: .caseInsensitive)
            for range in ranges {
                let match = KeywordMatch(
                    keyword: phrase,
                    range: range,
                    context: extractContext(from: text, around: range),
                    confidence: 0.8
                )
                exclusionMatches.append(match)
            }
        }
        
        let confidence = calculateOverallConfidence(
            keywordMatches: keywordMatches,
            exclusionMatches: exclusionMatches
        )
        
        return RuleBasedResult(
            hasArbitration: !keywordMatches.isEmpty && confidence > 0.5,
            confidence: confidence,
            keywordMatches: keywordMatches,
            exclusionMatches: exclusionMatches
        )
    }
    
    private func combineAnalysisResults(
        mlResult: Bool,
        mlConfidence: Double,
        ruleBasedResult: RuleBasedResult,
        originalText: String,
        analysisId: String
    ) -> ArbitrationAnalysisResult {
        
        // Weighted combination of ML and rule-based results
        let mlWeight = 0.7
        let ruleWeight = 0.3
        
        let combinedConfidence = (mlConfidence * mlWeight) + (ruleBasedResult.confidence * ruleWeight)
        let hasArbitration = combinedConfidence > configuration.confidenceThreshold
        
        return ArbitrationAnalysisResult(
            id: analysisId,
            hasArbitration: hasArbitration,
            confidence: combinedConfidence,
            keywordMatches: ruleBasedResult.keywordMatches,
            exclusionMatches: ruleBasedResult.exclusionMatches,
            riskLevel: calculateRiskLevel(confidence: combinedConfidence),
            recommendations: generateRecommendations(
                hasArbitration: hasArbitration,
                confidence: combinedConfidence,
                keywordMatches: ruleBasedResult.keywordMatches
            ),
            metadata: AnalysisMetadata(
                textLength: originalText.count,
                processingTime: 0, // Will be set later
                sdkVersion: "1.0.0",
                modelVersion: "1.0.0",
                analysisDate: Date()
            )
        )
    }
    
    // MARK: - Helper Methods
    
    private func countArbitrationKeywords(in text: String) -> Int {
        let keywords = ["arbitration", "arbitrator", "arbitral", "adr"]
        return keywords.reduce(0) { count, keyword in
            count + text.components(separatedBy: keyword).count - 1
        }
    }
    
    private func countLegalTerms(in text: String) -> Int {
        let terms = ["whereas", "hereby", "parties", "agreement", "contract"]
        return terms.reduce(0) { count, term in
            count + text.components(separatedBy: term).count - 1
        }
    }
    
    private func analyzeSentiment(_ text: String) -> Double {
        let tagger = NLTagger(tagSchemes: [.sentimentScore])
        tagger.string = text
        
        let sentiment = tagger.tag(at: text.startIndex, unit: .paragraph, scheme: .sentimentScore).0
        return Double(sentiment?.rawValue ?? "0") ?? 0.0
    }
    
    private func calculateComplexityScore(_ text: String) -> Double {
        let words = text.components(separatedBy: .whitespacesAndNewlines)
        let avgWordLength = words.map { $0.count }.reduce(0, +) / max(words.count, 1)
        let sentences = text.components(separatedBy: .punctuationCharacters)
        let avgSentenceLength = words.count / max(sentences.count, 1)
        
        return Double(avgWordLength + avgSentenceLength) / 10.0
    }
    
    private func calculateKeywordConfidence(_ keyword: String) -> Double {
        let weights: [String: Double] = [
            "arbitration": 0.9,
            "arbitrator": 0.8,
            "arbitral": 0.8,
            "binding arbitration": 0.95,
            "alternative dispute resolution": 0.7,
            "adr": 0.6
        ]
        return weights[keyword.lowercased()] ?? 0.5
    }
    
    private func extractContext(from text: String, around range: Range<String.Index>) -> String {
        let contextLength = 50
        let startIndex = max(text.startIndex, text.index(range.lowerBound, offsetBy: -contextLength, limitedBy: text.startIndex) ?? text.startIndex)
        let endIndex = min(text.endIndex, text.index(range.upperBound, offsetBy: contextLength, limitedBy: text.endIndex) ?? text.endIndex)
        return String(text[startIndex..<endIndex])
    }
    
    private func calculateOverallConfidence(keywordMatches: [KeywordMatch], exclusionMatches: [KeywordMatch]) -> Double {
        guard !keywordMatches.isEmpty else { return 0.0 }
        
        let keywordConfidence = keywordMatches.map { $0.confidence }.reduce(0, +) / Double(keywordMatches.count)
        let exclusionPenalty = exclusionMatches.isEmpty ? 0.0 : 0.3
        
        return max(0.0, min(1.0, keywordConfidence - exclusionPenalty))
    }
    
    private func calculateRiskLevel(confidence: Double) -> RiskLevel {
        switch confidence {
        case 0.8...:
            return .high
        case 0.5..<0.8:
            return .medium
        case 0.2..<0.5:
            return .low
        default:
            return .minimal
        }
    }
    
    private func generateRecommendations(hasArbitration: Bool, confidence: Double, keywordMatches: [KeywordMatch]) -> [String] {
        var recommendations: [String] = []
        
        if hasArbitration {
            recommendations.append("Arbitration clause detected. Review the specific terms and conditions.")
            
            if confidence > 0.8 {
                recommendations.append("High confidence detection. Consider legal consultation for detailed analysis.")
            } else {
                recommendations.append("Moderate confidence detection. Manual review recommended.")
            }
            
            if keywordMatches.contains(where: { $0.keyword.contains("binding") }) {
                recommendations.append("Binding arbitration clause found. This may limit legal recourse options.")
            }
        } else {
            recommendations.append("No clear arbitration clause detected in the analyzed text.")
            
            if confidence < 0.2 {
                recommendations.append("Consider reviewing the full document context for complete analysis.")
            }
        }
        
        return recommendations
    }
    
    private func extractTextFromDocument(_ data: Data, type: DocumentType) async throws -> String {
        // Implementation would depend on document type
        // For now, returning a placeholder
        throw ArbitrationError.unsupportedDocumentType("Document text extraction not yet implemented")
    }
    
    private func extractTextFromImage(_ image: UIImage) async throws -> String {
        return try await withCheckedThrowingContinuation { continuation in
            guard let cgImage = image.cgImage else {
                continuation.resume(throwing: ArbitrationError.invalidInput("Invalid image"))
                return
            }
            
            let request = VNRecognizeTextRequest { request, error in
                if let error = error {
                    continuation.resume(throwing: error)
                    return
                }
                
                let observations = request.results as? [VNRecognizedTextObservation] ?? []
                let recognizedText = observations.compactMap { observation in
                    observation.topCandidates(1).first?.string
                }.joined(separator: " ")
                
                continuation.resume(returning: recognizedText)
            }
            
            request.recognitionLevel = .accurate
            request.usesLanguageCorrection = true
            
            let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
            do {
                try handler.perform([request])
            } catch {
                continuation.resume(throwing: error)
            }
        }
    }
    
    @MainActor
    private func updateProgress(_ progress: Double) {
        analysisProgress = progress
    }
}

// MARK: - String Extension
private extension String {
    func ranges(of substring: String, options: CompareOptions = []) -> [Range<String.Index>] {
        var ranges: [Range<String.Index>] = []
        var searchStartIndex = self.startIndex
        
        while searchStartIndex < self.endIndex,
              let range = self.range(of: substring, options: options, range: searchStartIndex..<self.endIndex) {
            ranges.append(range)
            searchStartIndex = range.upperBound
        }
        
        return ranges
    }
}

File: sdks/android/src/main/java/com/arbitrationsdk/android/core/ArbitrationDetector.kt
package com.arbitrationsdk.android.core

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.arbitrationsdk.android.ml.MLModelManager
import com.arbitrationsdk.android.network.NetworkService
import com.arbitrationsdk.android.storage.StorageService
import com.arbitrationsdk.android.models.*
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import timber.log.Timber
import java.util.*
import javax.inject.Inject
import kotlin.time.Duration
import kotlin.time.DurationUnit
import kotlin.time.toDuration

/**
 * Main arbitration detection engine for Android
 * Provides comprehensive arbitration clause detection with ML and rule-based analysis
 */
@HiltViewModel
class ArbitrationDetector @Inject constructor(
    @ApplicationContext private val context: Context,
    private val networkService: NetworkService,
    private val storageService: StorageService,
    private val mlModelManager: MLModelManager,
    private val configuration: SDKConfiguration
) : ViewModel() {

    // State flows for reactive UI
    private val _isAnalyzing = MutableStateFlow(false)
    val isAnalyzing: StateFlow<Boolean> = _isAnalyzing.asStateFlow()

    private val _analysisProgress = MutableStateFlow(0f)
    val analysisProgress: StateFlow<Float> = _analysisProgress.asStateFlow()

    private val _lastAnalysisResult = MutableStateFlow<ArbitrationAnalysisResult?>(null)
    val lastAnalysisResult: StateFlow<ArbitrationAnalysisResult?> = _lastAnalysisResult.asStateFlow()

    private val _error = MutableSharedFlow<ArbitrationError>()
    val error: SharedFlow<ArbitrationError> = _error.asSharedFlow()

    // Analysis state
    private var currentAnalysisJob: kotlinx.coroutines.Job? = null

    init {
        // Initialize ML models
        viewModelScope.launch {
            try {
                mlModelManager.initializeModels()
                Timber.d("ML models initialized successfully")
            } catch (e: Exception) {
                Timber.e(e, "Failed to initialize ML models")
            }
        }
    }

    /**
     * Analyze text for arbitration clauses
     */
    suspend fun analyzeText(text: String): Result<ArbitrationAnalysisResult> {
        if (text.trim().isEmpty()) {
            return Result.failure(ArbitrationError.InvalidInput("Text cannot be empty"))
        }

        return withContext(Dispatchers.IO) {
            try {
                _isAnalyzing.value = true
                _analysisProgress.value = 0f

                val result = performAnalysis(text)
                
                // Store result locally
                storageService.saveAnalysisResult(result)
                
                _lastAnalysisResult.value = result
                Result.success(result)
            } catch (e: Exception) {
                val error = when (e) {
                    is ArbitrationError -> e
                    else -> ArbitrationError.AnalysisFailure(e.message ?: "Unknown error")
                }
                _error.emit(error)
                Result.failure(error)
            } finally {
                _isAnalyzing.value = false
                _analysisProgress.value = 0f
            }
        }
    }

    /**
     * Analyze document from byte array
     */
    suspend fun analyzeDocument(
        data: ByteArray,
        documentType: DocumentType
    ): Result<ArbitrationAnalysisResult> {
        return withContext(Dispatchers.IO) {
            try {
                val extractedText = extractTextFromDocument(data, documentType)
                analyzeText(extractedText)
            } catch (e: Exception) {
                Result.failure(ArbitrationError.UnsupportedDocumentType("Failed to extract text: ${e.message}"))
            }
        }
    }

    /**
     * Analyze image with OCR
     */
    suspend fun analyzeImage(imageData: ByteArray): Result<ArbitrationAnalysisResult> {
        return withContext(Dispatchers.IO) {
            try {
                val extractedText = extractTextFromImage(imageData)
                analyzeText(extractedText)
            } catch (e: Exception) {
                Result.failure(ArbitrationError.AnalysisFailure("OCR failed: ${e.message}"))
            }
        }
    }

    /**
     * Batch analyze multiple texts
     */
    suspend fun batchAnalyze(
        requests: List<AnalysisRequest>
    ): Result<BatchAnalysisResult> {
        if (requests.isEmpty()) {
            return Result.failure(ArbitrationError.InvalidInput("No requests provided"))
        }

        return withContext(Dispatchers.IO) {
            try {
                _isAnalyzing.value = true
                val batchId = UUID.randomUUID().toString()
                val results = mutableListOf<ArbitrationAnalysisResult>()
                val startTime = System.currentTimeMillis()

                requests.forEachIndexed { index, request ->
                    _analysisProgress.value = index.toFloat() / requests.size
                    
                    val result = performAnalysis(request.text)
                    results.add(result)
                    
                    // Store individual results
                    storageService.saveAnalysisResult(result)
                }

                val endTime = System.currentTimeMillis()
                val processingTime = (endTime - startTime).toDuration(DurationUnit.MILLISECONDS)

                val summary = BatchSummary(
                    totalDocuments = results.size,
                    documentsWithArbitration = results.count { it.hasArbitration },
                    averageConfidence = results.map { it.confidence }.average(),
                    highRiskDocuments = results.count { it.riskLevel == RiskLevel.HIGH },
                    processingErrors = 0
                )

                val batchResult = BatchAnalysisResult(
                    batchId = batchId,
                    results = results,
                    summary = summary,
                    processingTime = processingTime
                )

                Result.success(batchResult)
            } catch (e: Exception) {
                Result.failure(ArbitrationError.AnalysisFailure(e.message ?: "Batch analysis failed"))
            } finally {
                _isAnalyzing.value = false
                _analysisProgress.value = 0f
            }
        }
    }

    /**
     * Cancel ongoing analysis
     */
    fun cancelAnalysis() {
        currentAnalysisJob?.cancel()
        _isAnalyzing.value = false
        _analysisProgress.value = 0f
    }

    /**
     * Get analysis history
     */
    fun getAnalysisHistory(
        limit: Int = 50,
        offset: Int = 0,
        sortBy: SortOption = SortOption.DATE_DESCENDING,
        filter: ResultFilter? = null
    ): Flow<List<ArbitrationAnalysisResult>> {
        return storageService.getAllAnalysisResults(limit, offset, sortBy, filter)
    }

    // Private implementation methods

    private suspend fun performAnalysis(text: String): ArbitrationAnalysisResult {
        val analysisId = UUID.randomUUID().toString()
        val startTime = System.currentTimeMillis()

        updateProgress(0.1f)

        // Step 1: Preprocessing
        val preprocessedText = preprocessText(text)
        updateProgress(0.2f)

        // Step 2: Feature extraction
        val features = extractFeatures(preprocessedText)
        updateProgress(0.4f)

        // Step 3: ML Model prediction
        val (mlResult, mlConfidence) = if (mlModelManager.isModelAvailable()) {
            mlModelManager.predict(features)
        } else {
            Pair(false, 0.0)
        }
        updateProgress(0.6f)

        // Step 4: Rule-based analysis
        val ruleBasedResult = performRuleBasedAnalysis(preprocessedText)
        updateProgress(0.8f)

        // Step 5: Combine results
        val finalResult = combineAnalysisResults(
            mlResult = mlResult,
            mlConfidence = mlConfidence,
            ruleBasedResult = ruleBasedResult,
            originalText = text,
            analysisId = analysisId
        )

        updateProgress(1.0f)

        val endTime = System.currentTimeMillis()
        val processingTime = (endTime - startTime).toDuration(DurationUnit.MILLISECONDS)
        
        return finalResult.copy(
            metadata = finalResult.metadata.copy(
                processingTime = processingTime
            )
        )
    }

    private fun preprocessText(text: String): String {
        return text.lowercase(Locale.getDefault())
            .replace(Regex("\\s+"), " ")
            .trim()
    }

    private fun extractFeatures(text: String): Map<String, Any> {
        val words = text.split(Regex("\\s+"))
        
        return mapOf(
            "text_length" to text.length,
            "word_count" to words.size,
            "arbitration_keywords" to countArbitrationKeywords(text),
            "legal_terms" to countLegalTerms(text),
            "sentence_count" to text.split(Regex("[.!?]+")).size,
            "avg_word_length" to if (words.isNotEmpty()) words.map { it.length }.average() else 0.0,
            "complexity_score" to calculateComplexityScore(text)
        )
    }

    private fun countArbitrationKeywords(text: String): Int {
        val keywords = listOf(
            "arbitration", "arbitrator", "arbitral", "adr",
            "binding arbitration", "alternative dispute resolution"
        )
        return keywords.sumOf { keyword ->
            text.split(keyword, ignoreCase = true).size - 1
        }
    }

    private fun countLegalTerms(text: String): Int {
        val terms = listOf(
            "whereas", "hereby", "parties", "agreement", "contract",
            "terms", "conditions", "jurisdiction", "governed"
        )
        return terms.sumOf { term ->
            text.split(term, ignoreCase = true).size - 1
        }
    }

    private fun calculateComplexityScore(text: String): Double {
        val words = text.split(Regex("\\s+"))
        val avgWordLength = if (words.isNotEmpty()) words.map { it.length }.average() else 0.0
        val sentences = text.split(Regex("[.!?]+"))
        val avgSentenceLength = if (sentences.isNotEmpty()) words.size.toDouble() / sentences.size else 0.0
        
        return (avgWordLength + avgSentenceLength) / 10.0
    }

    private fun performRuleBasedAnalysis(text: String): RuleBasedResult {
        val arbitrationKeywords = listOf(
            "arbitration", "arbitrator", "arbitral", "binding arbitration",
            "alternative dispute resolution", "adr", "arbitration clause",
            "dispute resolution", "binding resolution", "final and binding"
        )

        val exclusionPhrases = listOf(
            "court proceedings", "litigation", "judicial review",
            "class action", "jury trial", "court of law"
        )

        val keywordMatches = mutableListOf<KeywordMatch>()
        val exclusionMatches = mutableListOf<KeywordMatch>()

        // Find arbitration keywords
        arbitrationKeywords.forEach { keyword ->
            val regex = Regex("\\b${Regex.escape(keyword)}\\b", RegexOption.IGNORE_CASE)
            regex.findAll(text).forEach { match ->
                val keywordMatch = KeywordMatch(
                    keyword = keyword,
                    range = "${match.range.first}:${match.range.last}",
                    context = extractContext(text, match.range),
                    confidence = calculateKeywordConfidence(keyword)
                )
                keywordMatches.add(keywordMatch)
            }
        }

        // Find exclusion phrases
        exclusionPhrases.forEach { phrase ->
            val regex = Regex("\\b${Regex.escape(phrase)}\\b", RegexOption.IGNORE_CASE)
            regex.findAll(text).forEach { match ->
                val exclusionMatch = KeywordMatch(
                    keyword = phrase,
                    range = "${match.range.first}:${match.range.last}",
                    context = extractContext(text, match.range),
                    confidence = 0.8
                )
                exclusionMatches.add(exclusionMatch)
            }
        }

        val confidence = calculateOverallConfidence(keywordMatches, exclusionMatches)

        return RuleBasedResult(
            hasArbitration = keywordMatches.isNotEmpty() && confidence > 0.5,
            confidence = confidence,
            keywordMatches = keywordMatches,
            exclusionMatches = exclusionMatches
        )
    }

    private fun extractContext(text: String, range: IntRange): String {
        val contextLength = 50
        val start = maxOf(0, range.first - contextLength)
        val end = minOf(text.length, range.last + contextLength)
        return text.substring(start, end)
    }

    private fun calculateKeywordConfidence(keyword: String): Double {
        val weights = mapOf(
            "arbitration" to 0.9,
            "arbitrator" to 0.8,
            "arbitral" to 0.8,
            "binding arbitration" to 0.95,
            "alternative dispute resolution" to 0.7,
            "adr" to 0.6
        )
        return weights[keyword.lowercase()] ?: 0.5
    }

    private fun calculateOverallConfidence(
        keywordMatches: List<KeywordMatch>,
        exclusionMatches: List<KeywordMatch>
    ): Double {
        if (keywordMatches.isEmpty()) return 0.0

        val keywordConfidence = keywordMatches.map { it.confidence }.average()
        val exclusionPenalty = if (exclusionMatches.isNotEmpty()) 0.3 else 0.0

        return maxOf(0.0, minOf(1.0, keywordConfidence - exclusionPenalty))
    }

    private fun combineAnalysisResults(
        mlResult: Boolean,
        mlConfidence: Double,
        ruleBasedResult: RuleBasedResult,
        originalText: String,
        analysisId: String
    ): ArbitrationAnalysisResult {
        // Weighted combination of ML and rule-based results
        val mlWeight = 0.7
        val ruleWeight = 0.3

        val combinedConfidence = (mlConfidence * mlWeight) + (ruleBasedResult.confidence * ruleWeight)
        val hasArbitration = combinedConfidence > configuration.confidenceThreshold

        return ArbitrationAnalysisResult(
            id = analysisId,
            hasArbitration = hasArbitration,
            confidence = combinedConfidence,
            keywordMatches = ruleBasedResult.keywordMatches,
            exclusionMatches = ruleBasedResult.exclusionMatches,
            riskLevel = calculateRiskLevel(combinedConfidence),
            recommendations = generateRecommendations(hasArbitration, combinedConfidence, ruleBasedResult.keywordMatches),
            metadata = AnalysisMetadata(
                textLength = originalText.length,
                processingTime = Duration.ZERO, // Will be set later
                sdkVersion = "1.0.0",
                modelVersion = "1.0.0",
                analysisDate = Date()
            )
        )
    }

    private fun calculateRiskLevel(confidence: Double): RiskLevel {
        return when {
            confidence >= 0.8 -> RiskLevel.HIGH
            confidence >= 0.5 -> RiskLevel.MEDIUM
            confidence >= 0.2 -> RiskLevel.LOW
            else -> RiskLevel.MINIMAL
        }
    }

    private fun generateRecommendations(
        hasArbitration: Boolean,
        confidence: Double,
        keywordMatches: List<KeywordMatch>
    ): List<String> {
        val recommendations = mutableListOf<String>()

        if (hasArbitration) {
            recommendations.add("Arbitration clause detected. Review the specific terms and conditions.")

            if (confidence > 0.8) {
                recommendations.add("High confidence detection. Consider legal consultation for detailed analysis.")
            } else {
                recommendations.add("Moderate confidence detection. Manual review recommended.")
            }

            if (keywordMatches.any { it.keyword.contains("binding", ignoreCase = true) }) {
                recommendations.add("Binding arbitration clause found. This may limit legal recourse options.")
            }
        } else {
            recommendations.add("No clear arbitration clause detected in the analyzed text.")

            if (confidence < 0.2) {
                recommendations.add("Consider reviewing the full document context for complete analysis.")
            }
        }

        return recommendations
    }

    private suspend fun extractTextFromDocument(data: ByteArray, type: DocumentType): String {
        // Placeholder for document text extraction
        // In a real implementation, you would use libraries like Apache Tika or specific parsers
        throw ArbitrationError.UnsupportedDocumentType("Document text extraction not yet implemented")
    }

    private suspend fun extractTextFromImage(imageData: ByteArray): String {
        // Placeholder for OCR implementation
        // In a real implementation, you would use ML Kit Text Recognition or Tesseract
        throw ArbitrationError.AnalysisFailure("OCR text extraction not yet implemented")
    }

    private suspend fun updateProgress(progress: Float) {
        _analysisProgress.value = progress
    }

    override fun onCleared() {
        super.onCleared()
        cancelAnalysis()
    }
}